File: homework1/scripts/goal_pub.py
#!/usr/bin/env python
import rospy
from turtlesim.msg import Pose
from math import sqrt
from std_srvs.srv import Empty
class position:
    
    def __init__(self,x,y):
        self.x = x
        self.y = y
    

class collect_max_coins:


    def __init__(self,r,c):
        self.r = r+1
        self.c = c+1
        self.table = [[0 for i in range(c+1)] for j in range(r+1)]
        self.coins_distr = [[0 for i in range(c+1)] for j in range(r+1)]
        self.positions = []


        rospy.init_node('goal_pub',anonymous=True)

        self.goal_pub = rospy.Publisher('/goal_info',Pose,queue_size=10)
        self.pose_sub = rospy.Subscriber('/turtle1/pose',Pose,self.update_pose)
        self.pose = Pose()
        self.rate = rospy.Rate(10)

        


    def update_pose(self, data):
       
        self.pose = data
        self.pose.x = round(self.pose.x, 4)
        self.pose.y = round(self.pose.y, 4)
    
    def get_coins_distr(self):
        for i in range(self.r):
            l = raw_input()
            line = l.split(' ')
            for j in range(self.c):
                self.coins_distr[i][j] = int(line[j])
                



    def fill_table(self):
        for i in range(1,self.r):
            for j in range(1,self.c):
                if self.table[i - 1][j]>self.table[i][j-1]:
                    self.table[i][j] += self.coins_distr[i][j] + self.table[i-1][j]
                else:
                    self.table[i][j] += self.coins_distr[i][j] + self.table[i][j-1]
                    

    def show(self):
        for i in self.table:
            print(i)

    def cmp(self,p1,p2):
        if(p1.y != p2.y): return p2.y - p1.y
        else: return p1.x - p2.x


    def get_path(self):
        i = self.r - 1
        j = self.c - 1
        while(j != 0 and i != 0):
            
            p = position(j,self.r - i)
            
            self.positions.append(p)

            if j == 1: i = i - 1
                
            elif i == 1: j =j - 1
                
            elif self.table[i-1][j]>self.table[i][j-1]: i = i - 1

            else: j = j - 1

        self.positions = sorted(self.positions,cmp=self.cmp)

        for i in self.positions:
            print("x: ",i.x," y: ",i.y)
                
    def start(self):

        pose = Pose()
        rospy.wait_for_service('/clear')
        clin = rospy.ServiceProxy('/clear',Empty)
        for p in self.positions:
            pose.x = p.x
            pose.y = p.y
            mx = self.pose.x - p.x
            my = self.pose.y - p.y
            distance = sqrt(mx*mx + my*my)

            while distance>0.02 and not rospy.is_shutdown():
                self.goal_pub.publish(pose)
                mx = self.pose.x - p.x
                my = self.pose.y - p.y
                distance = sqrt(mx*mx + my*my)
        
            if p.x == 1 and p.y == 5 : clin()
            
            self.coins_distr[self.r - p.y][p.x] = 8
            for i in self.coins_distr:
                print(i)
            
        

        self.rate.sleep()

            

            



if __name__ == "__main__":
    c = collect_max_coins(5,6)
    c.get_coins_distr()
    c.fill_table()
    
    c.get_path()
    c.start()
    
    pass

File: homework1/src/goal_pub.cpp
#include<ros/ros.h>
#include<geometry_msgs/Twist.h>
#include<cmath>
#include"turtlesim/Pose.h"
#include<std_srvs/Empty.h>
#include<iostream>
#include<algorithm>

using namespace std;


class collect_max_coins{
    
public:

    int matrix[6][7];//6x7 simplify
    int* a[6];
    struct position
    {
        int x,y;
        /* data */
    }positions[100];
    
    int pindex;

    collect_max_coins(){
        for(int i = 0;i<6;i++){
            for(int j = 0;j<7;j++)matrix[i][j] = 0;
        }
    }

    void fill_table(int* a[]){
        for(int i = 1;i<6;i++){
            for(int j = 1;j<7;j++){
                int max = matrix[i-1][j] > matrix[i][j-1] ? matrix[i-1][j] : matrix[i][j-1];

                matrix[i][j] = max + a[i][j];
            }
        }   
        
        for(int i = 0;i<6;i++){
            for(int j = 0;j<7;j++)cout<<matrix[i][j]<<" ";
            cout<<endl;
        }
        
    }

    void get_coins_distr(){
        int pos;
        
        for(int i = 0;i<7;i++)a[i] = new int[7]; 
        for(int i = 0;i<6;i++){
            for(int j = 0;j<7;j++){
                cin>>pos;
                a[i][j] = pos;
            }
        }cout<<endl;
        fill_table(a);
    }

    static int cmp1(position a,position b){
        if(a.y != b.y)return a.y>b.y;
        return a.x<b.x;
    }

    // int cmp2(position a,position b){
    //     return a.y<b.y;
    // }



    void get_path(){
        int i = 5;
        int j = 6;
        pindex = 0;
        position p; 
        // p.x = 1; p.y = 6-1;
        // positions[pindex++] = p;
        while(true){
            p.x = j; p.y = 6-i;
            positions[pindex++] = p;
           
            if(i == 1)j--;
            else if(j == 1)i--;
            else if(matrix[i-1][j]>matrix[i][j-1]) i--;
            else j--;

            if(i == 0 || j == 0)break;
                    
            
                    
                
        }
        // p.x = j;p.y =6-i;
        // positions[pindex++] = p;

        sort(positions,positions+pindex,cmp1);

        for(int i = 0;i<pindex;i++){
            cout<<positions[i].x<<" "<<positions[i].y<<endl;
        }

    }
    
        

};

turtlesim::Pose* p = new turtlesim::Pose();

void poseCallback(const turtlesim::Pose::ConstPtr& msg){

        

    p->x = msg->x;
    p->y = msg->y;

            



}

int main(int argc, char **argv)
{
    ros::init(argc,argv,"goal_pub");
    ros::NodeHandle n;
    collect_max_coins c;
    ros::Publisher goal_pub = n.advertise<turtlesim::Pose>("/goal_info",10);
    ros::Subscriber pose_sub = n.subscribe("/turtle1/pose",10,poseCallback);
    
    ros::service::waitForService("/clear");
    ros::ServiceClient clin = n.serviceClient<std_srvs::Empty>("/clear");
    
    std_srvs::Empty srv;

    ros::Rate loop_rate(1);
    loop_rate.sleep();
    ros::spinOnce();
    
    
    
    
    
    
    c.get_coins_distr();
    c.get_path();
    
    

    
        turtlesim::Pose goal_info;
        
        
        for(int i = 0;i<c.pindex && ros::ok();i++){
            ros::spinOnce();
            

            int x = c.positions[i].x;
            int y = c.positions[i].y;
            goal_info.x = x;
            goal_info.y = y;
            
            

            double xm = p->x - x;
            double ym = p->y - y;
            double distance = sqrt(xm*xm+ym*ym);

            while(distance >= 0.02 && ros::ok){

                goal_pub.publish(goal_info);
                
                
                ros::spinOnce();
                
                xm = p->x - x;
                ym = p->y - y;
                distance = sqrt(xm*xm+ym*ym);
            }

            if(x == 1 && y == 5)clin.call(srv);
            c.a[6-y][x] = 8;
            for(int i = 0;i<6;i++){
                for(int j = 0;j<7;j++)cout<<c.a[i][j]<<" ";
                cout<<endl;
            }
            cout<<endl;


            loop_rate.sleep();

            
        }
    

        
    
}